common/persistence/elasticsearch/visibility_store.go

```diff
@@ -1076,19 +1081,48 @@ func processAllValuesForKey(dsl *fastjson.Value, keyFilter func(k string) bool,
 }
 
 func timeKeyFilter(key string) bool {
-	return timeKeys[key]
+	_, ok := timeKeys[key]
+	return ok
+}
+
+func timeProcessFunc(_ *fastjson.Object, _ string, value *fastjson.Value) error {
+	return processAllValuesForKey(
+		value,
+		func(key string) bool {
+			_, ok := rangeKeys[key]
+			return ok
+		},
+		func(obj *fastjson.Object, key string, v *fastjson.Value) error {
+			timeStr := string(v.GetStringBytes())
+
+			// To support dates passed as int64 "nanoseconds since epoch".
+			if nanos, err := strconv.ParseInt(timeStr, 10, 64); err == nil {
+				obj.Set(key, fastjson.MustParse(fmt.Sprintf(`"%s"`, time.Unix(0, nanos).UTC().Format(time.RFC3339Nano))))
+			}
+			return nil
+		})
 }
 
-func timeProcessFunc(obj *fastjson.Object, key string, value *fastjson.Value) error {
-	return processAllValuesForKey(value, func(key string) bool {
-		return rangeKeys[key]
-	}, func(obj *fastjson.Object, key string, v *fastjson.Value) error {
-		timeStr := string(v.GetStringBytes())
+func statusKeyFilter(key string) bool {
+	return key == searchattribute.ExecutionStatus
+}
 
-		// To support dates passed as "nanoseconds since epoch".
-		if nanos, err := strconv.ParseInt(timeStr, 10, 64); err == nil {
-			obj.Set(key, fastjson.MustParse(fmt.Sprintf(`"%s"`, time.Unix(0, nanos).UTC().Format(time.RFC3339Nano))))
-		}
-		return nil
-	})
+func statusProcessFunc(_ *fastjson.Object, _ string, value *fastjson.Value) error {
+	return processAllValuesForKey(
+		value,
+		func(key string) bool {
+			_, ok := exactMatchKeys[key]
+			return ok
+		},
+		func(obj *fastjson.Object, key string, v *fastjson.Value) error {
+			statusStr := string(v.GetStringBytes())
+
+			// To support statuses passed as integers for backward compatibility.
+			// Might be removed one day (added 6/15/21).
+			if statusInt, err := strconv.ParseInt(statusStr, 10, 32); err == nil {
+				statusStr = enumspb.WorkflowExecutionStatus_name[int32(statusInt)]
+				obj.Set(key, fastjson.MustParse(fmt.Sprintf(`"%s"`, statusStr)))
+			}
+			return nil
```

Comment:
I'm confused. Why we need to use processAllValuesForKey in statusProcessFunc? Isn't statusProcessFunc call by processAllValuesForKey (at line 588) which already did the filtering.

Reply:
Agree, this is very confusing code. I don't want to change much in current logic (`statusProcessFunc` is just a copy of `timeProcessFunc`) but I left TODO for myself to revisit all this JSON post processing with separate PR. Basically the logic is followting:
1. Convert SQL-ish query to ES JSON query. This is done by `elasticsql` package. Queries like `ExecutionStatus="Running"` are converted to: `{"query":{"bool":{"must":[{"match_phrase":{"ExecutionStatus":{"query":"Running"}}}]}},"from":0,"size":20}`
2. Then `fastjson` parse this JSON and substitute some values.
3. `statusKeyFilter` catch `ExecutionStatus` key and sends its value `{"query":"Running"}` to the `statusProcessFunc`.
4. It treats this value as regular JSON and calls `processAllValuesForKey` for it.
5. 1113-1116 func catched `query` key and call this handler passing its value `"Running"`.
6. In case of `string` it is just ignored but if it is a `int`, it gets converted to `string` and set back to `obj`.
7. Final query is sent to Elasticsearch and it perform search.

Actually, Elasticsearch has its own SQL support and starting with version 6.3 it is in open source: https://www.elastic.co/what-is/open-x-pack. So probably all this code can be just replaced with simple query augmentor.

Reply:
Thanks for the explaining. Could you please paste those steps as comments in code. And add a TODO or task to replace all those with the ES SQL query.